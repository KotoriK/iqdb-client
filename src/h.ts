export interface IQDBClientOptions {
    baseDomain: string,
    simlarityPass: number
    userAgent:string
}
/**'www'search all libs */
export type IQDB_SEARCH_LIBRARY_2D = 'danbooru' | 'konachan' | 'yandere' | 'gelbooru' | 'sankaku' | 'e-shuushuu' | 'zerochan' | 'anime-pictures' | 'www'
/**cosplay, asian models, idols.'3d' search all libs.*/
export type IQDB_SEARCH_LIBRARY_3D = '3d' | '3dbooru' | 'idol'
export interface IQDB_SEARCH_OPTIONS_GENERAL {
    forcegray?: boolean
    fileName?:string
}
export type IQDB_SEARCH_OPTIONS_2D = ({
    lib: Exclude<IQDB_SEARCH_LIBRARY_2D, 'www'>,
    libs?:undefined

} | {
    lib: 'www',
    libs?: Array<IQDBLibs_2D>
}) & IQDB_SEARCH_OPTIONS_GENERAL
export type IQDB_SEARCH_OPTIONS_3D = ({
    lib: Exclude<IQDB_SEARCH_LIBRARY_3D, '3d'>,
    libs?:undefined
} | {
    lib: '3d',
    libs?: Array<IQDBLibs_3D>
}) & IQDB_SEARCH_OPTIONS_GENERAL
export type IQDB_SEARCH_OPTIONS_ALL = IQDB_SEARCH_OPTIONS_2D | IQDB_SEARCH_OPTIONS_3D

export enum IQDBLibs_2D {
    danbooru = 1,
    konachan = 2,
    'yande.re' = 3,
    gelbooru = 4,
    'sankaku channel' = 5,
    'e-shuushuu' = 6,
    zerochan = 11,
    'anime-picture' = 13
}
export enum IQDBLibs_3D {
    '3dbooru' = 7,
    'idol' = 9
}
