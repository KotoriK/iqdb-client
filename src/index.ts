/**
 * iqdb-client
 * @description iqdb.org api client for Node.js.
 * @version 1.0.4
 * @author KotoriK
 * @license MIT
 */
import searchPic from './api'
export { IQDBSearchResultItem, Size, IQDB_OPTIONS, setIQDBOptions } from './api'
export { searchPic }
export * from './h'