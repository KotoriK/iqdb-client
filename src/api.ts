import { parseNum, parseSizeAndType, randomFileName } from "./util"
import { load } from 'cheerio'
import fetch from 'node-fetch'
import { IQDB_SEARCH_OPTIONS_ALL, IQDBClientOptions, IQDBLibs_2D, IQDBLibs_3D, IQDB_RESULT_TYPE } from "./h"
import FormData from 'form-data'
import { Readable } from 'stream'
export interface Size {
    width: number
    height: number
}
export type IQDBSearchResultItem = {
    head: string,
    img: string,
    name?: string,
    sourceUrl?: string,
    source?: string[],
    similarity?: number
} & ({
    size?: Size,
    type?: IQDB_RESULT_TYPE
} | { sizeAndType: string })
export function setIQDBOptions(newOption: IQDBClientOptions) {
    IQDB_OPTIONS = newOption
}
export let IQDB_OPTIONS: IQDBClientOptions = {
    baseDomain: 'iqdb.org',
    simlarityPass: 0.6,
    userAgent: 'node-fetch/1.0 (+https://github.com/bitinn/node-fetch)',
}
const _addToForm = (form: FormData, array: Array<IQDBLibs_2D | IQDBLibs_3D>) => array.forEach(lib => { form.append('service[]', lib) })

/**
 * 
 * @param body 服务器返回的body
 * @param noSource 指示结果中是否应该有source字段
 * @returns
 */
export function _parseResult(body: string, noSource?: boolean) {
    const $ = load(body)
    let ok = false
    const err = $('.err')
    if (err.length > 0) {
        return {
            /**是否找到满足相似度的结果 */
            ok,
            /**是否发生错误 */
            err: err.text()
        }
    }
    const data: Array<IQDBSearchResultItem> = $('#pages').children('div').toArray()
        .map(page => {
            const rows = $(page).find('tr')
            const head = $(rows[0]).text()
            if (head == 'Your image') {
                const sizeAndType = parseSizeAndType($(rows[3]).text())
                return {
                    head,
                    img: $(rows[1]).find('img').attr('src'),
                    name: $(rows[2]).text(),
                    ...typeof sizeAndType == 'object' ? sizeAndType : { sizeAndType }
                }
            } else if (head == 'No relevant matches') {
                return
            } else {
                let similarity: number | string, sizeAndType: string | { size: { width: number; height: number }; type: string }, source: string[]
                if (noSource) {
                    similarity = parseNum($(rows[3]).text())
                    sizeAndType = parseSizeAndType($(rows[2]).text())
                } else {
                    similarity = parseNum($(rows[4]).text())
                    sizeAndType = parseSizeAndType($(rows[3]).text())
                    source = $(rows[2]).text().split(' ')
                }
                if (similarity >= IQDB_OPTIONS.simlarityPass) ok = true
                const imgBox = $(rows[1]).find('a')

                return {
                    head,
                    sourceUrl: imgBox.attr('href'),
                    img: imgBox.find('img').attr('src'),
                    ...typeof sizeAndType == 'object' ? sizeAndType : { sizeAndType },
                    source
                }
            }
        })
        .filter(item => item != undefined)
    return {
        /**是否找到满足相似度的结果 */
        ok,
        /**返回数据 */
        data
    }
}
export default async function searchPic(pic: string | Buffer | Readable, { lib, forcegray, libs, fileName }: IQDB_SEARCH_OPTIONS_ALL) {
    const isMultiLib = lib == 'www' || lib == '3d'
    const form = new FormData()
    if (typeof pic == 'string') { form.append('url', pic) }
    else if (pic instanceof Buffer || pic instanceof Readable) { form.append('file', pic, { filename: fileName ? fileName : randomFileName() }) }
    else {
        throw new TypeError('expect string | Buffer | Readable')
    }
    if (isMultiLib && libs) _addToForm(form, libs)
    if (forcegray) form.append('forcegray', 'true')
    const resp = await fetch(`https://${lib}.${IQDB_OPTIONS.baseDomain}`,
        {
            method: 'POST',
            body: form,
            headers: {
                ...form.getHeaders(),
                'User-Agent': IQDB_OPTIONS.userAgent
            },
            ...IQDB_OPTIONS.fetchOptions
        })
    if (resp.ok) {
        return _parseResult(await resp.text(), !isMultiLib)
    } else {
        return {
            ok: false,
            err: 'HTTP ' + resp.status
        }
    }
}