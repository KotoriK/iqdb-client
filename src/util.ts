import { Size } from "./api"
import { IQDB_RESULT_TYPE } from "./h"

export function parseSimilarity(txt: string): number | null {
    const result = txt.match(/(-?\d+\.?\d*)(%?)/)
    if (result) {
        return (result[2] == undefined || result[2] == '') ? parseFloat(result[1]) : parseFloat(result[1]) / 100
    } else {
        return null//couldn't parse similarity
    }
}
export function parseSizeAndType(txt: string): { size: Size, type: IQDB_RESULT_TYPE } | string {
    const result = txt.match(/(\d+)Ã—(\d+)(?: \[([A-z]*)\])?/)
    if (result) {
        return {
            size: {
                width: parseInt(result[1]),
                height: parseInt(result[2])
            },
            type: result[3]
        }
    } else {
        return txt
    }
}
const availCharset = '0123456789abcdefghijklmnopqrstuvwxyz@ABCDEFGHIJKLMNOPQRSTUVWXYZ_'
const availCharsetLen = availCharset.length
export function getRandomName() {
    const length = 5 + (Math.random() * 10 | 0)
    let str = ''
    for (let i = 0; i < length; i++) {
        str += availCharset[Math.random() * availCharsetLen | 0]
    }
    return str + '.jpg'
}
