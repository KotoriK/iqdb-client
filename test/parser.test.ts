import { _parseResult } from '../src/api'
import fs from 'fs/promises'
import { resolve } from 'path'
test('parser', async() => {
    const parseResult = fs.readFile(resolve(__dirname, '../test/response.html'), { encoding: 'utf-8' })
        .then(txt => _parseResult(txt))
    const shouldBe = fs.readFile(resolve(__dirname, '../test/response.json'), { encoding: 'utf-8' })
        .then(txt => JSON.parse(txt))
    expect(await parseResult).toEqual(await shouldBe)
})
