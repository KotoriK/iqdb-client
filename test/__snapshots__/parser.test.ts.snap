// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseResult parses response1 1`] = `
{
  "data": [
    {
      "head": "Your image",
      "img": "/thu/thu_114514.jpg",
      "name": "84035784_p2.jpg",
      "similarity": null,
      "size": {
        "height": 2459,
        "width": 1703,
      },
      "type": undefined,
    },
    {
      "head": "Best match",
      "img": "/danbooru/1/f/8/1f8ff3c560a0689e795938138dac7b1f.jpg",
      "similarity": 0.96,
      "size": {
        "height": 2459,
        "width": 1703,
      },
      "source": [
        "Danbooru",
        "Gelbooru",
      ],
      "sourceUrl": "//danbooru.donmai.us/posts/4076714",
      "type": "Safe",
    },
    {
      "head": "Additional match",
      "img": "/moe.imouto/8/0/1/801df5f665e61e6f87eb85431f2ca2a1.jpg",
      "similarity": 0.92,
      "size": {
        "height": 2459,
        "width": 1703,
      },
      "source": [
        "yande.re",
      ],
      "sourceUrl": "https://yande.re/post/show/678391",
      "type": "Safe",
    },
  ],
  "ok": true,
  "service": [
    1,
    2,
    3,
    4,
    5,
    6,
    11,
    13,
  ],
}
`;

exports[`parseResult parses response2 1`] = `
{
  "data": [
    {
      "head": "Your image",
      "img": "/thu/thu_63ea49f3.jpg",
      "name": "i10yn9O72A33.jpg",
      "similarity": null,
      "size": {
        "height": 3300,
        "width": 2285,
      },
      "type": undefined,
    },
    {
      "head": "Best match",
      "img": "/danbooru/1/e/a/1eae91afbfc9610ed741766d3f1c6aa4.jpg",
      "similarity": 0.95,
      "size": {
        "height": 3300,
        "width": 2285,
      },
      "source": [
        "Danbooru",
        "Gelbooru",
      ],
      "sourceUrl": "//danbooru.donmai.us/posts/5524659",
      "type": "Ero",
    },
    {
      "head": "Additional match",
      "img": "/moe.imouto/1/e/a/1eae91afbfc9610ed741766d3f1c6aa4.jpg",
      "similarity": 0.95,
      "size": {
        "height": 3300,
        "width": 2285,
      },
      "source": [
        "yande.re",
      ],
      "sourceUrl": "https://yande.re/post/show/1000890",
      "type": "Ero",
    },
    {
      "head": "Additional match",
      "img": "/moe.imouto/4/9/f/49f8d06db9296fd9a887964e64aa990b.jpg",
      "similarity": 0.61,
      "size": {
        "height": 4096,
        "width": 2673,
      },
      "source": [
        "yande.re",
      ],
      "sourceUrl": "https://yande.re/post/show/1001188",
      "type": "Ero",
    },
  ],
  "ok": true,
  "service": [
    1,
    2,
    3,
    4,
    5,
    6,
    11,
    13,
  ],
}
`;
